eclair-core/src/main/scala/fr/acinq/eclair/router/RouteCalculation.scala

```diff
@@ -38,14 +38,16 @@ object RouteCalculation {
   def finalizeRoute(d: Data, fr: FinalizeRoute)(implicit ctx: ActorContext, log: LoggingAdapter): Data = {
     implicit val sender: ActorRef = ctx.self // necessary to preserve origin when sending messages to other actors
 
-    // NB: using a capacity of 0 msat will impact the path-finding algorithm. However here we don't run any path-finding, so it's ok.
-    val assistedChannels: Map[ShortChannelId, AssistedChannel] = fr.assistedRoutes.flatMap(toAssistedChannels(_, fr.hops.last, 0 msat)).toMap
-    val extraEdges = assistedChannels.values.map(ac => GraphEdge(ChannelDesc(ac.extraHop.shortChannelId, ac.extraHop.nodeId, ac.nextNodeId), toFakeUpdate(ac.extraHop, ac.htlcMaximum))).toSet
+    val assistedChannels: Map[ShortChannelId, AssistedChannel] = fr.assistedRoutes.flatMap(toAssistedChannels(_, fr.hops.last, fr.amount)).toMap
+    val extraEdges = assistedChannels.values.map(ac =>
+      GraphEdge(ChannelDesc(ac.extraHop.shortChannelId, ac.extraHop.nodeId, ac.nextNodeId), toFakeUpdate(ac.extraHop, ac.htlcMaximum), htlcMaxToCapacity(ac.htlcMaximum), Some(ac.htlcMaximum))
+    ).toSet
     val g = extraEdges.foldLeft(d.graph) { case (g: DirectedGraph, e: GraphEdge) => g.addEdge(e) }
     // split into sublists [(a,b),(b,c), ...] then get the edges between each of those pairs
     fr.hops.sliding(2).map { case List(v1, v2) => g.getEdgesBetween(v1, v2) }.toList match {
       case edges if edges.nonEmpty && edges.forall(_.nonEmpty) =>
-        val selectedEdges = edges.map(_.maxBy(_.update.htlcMaximumMsat.getOrElse(0 msat))) // select the largest edge
+        // select the largest edge (using balance when available, otherwise capacity).
+        val selectedEdges = edges.map(es => es.maxBy(e => e.balance_opt.getOrElse(e.capacity.toMilliSatoshi)))
```

Comment:
Is there the chance that we mix the largest by `capacity` and by `balance`? The `balance` should be available on all local channels but if we end up having a `None` then we'd effectively compare balance and capacity which is wrong, 
for example in the following situation we'd pick the wrong channel:
1. local_channel_AB: _toLocal_: 1000 sat, _toRemote_: 50000 sat, _capacity_: 50000 sat
2. local_channel_CB: _toLocal_: 0 sat, _toRemote_: 60000 sat,  _capacity_: 60000 sat

Maybe we can add a test to make sure this doesn't happen.

Reply:
There is a small window where a race is possible and the router still has a `None` for some of the local channels (after a node restart). But I don't think it's an issue. We've lived this far with only looking at the capacity without too many issues. Looking at the balance when available is already a net improvement over that. And if that payment attempt fails, the retry will very likely not hit the race condition and will have all balances available.

Note that in your example, the previous code would have failed to route all the time. Now it has a chance to route once the router has the balance information available. It's just a matter of waiting a bit to have the information propagate. I don't think it's worth adding more complexity here for such an edge case.